Add encoding_len to ProtobufField

bryfox opened the pull request:
This builds on #324, and is a follow-up to #396.

The primary change is to add a required `encoded_len` method to the ProtobufField trait. This provides the length of encoded data to `Encode::encode`, which will now return an error if the remaining buffer is too short to hold the written data.

This also adds a compiler check for unit-only enums; any variant with fields will now provide a nicer error message.

This also updates the documentation, and notes the protobuf limitation around enum values.

This also fixes the encoding of tags, which were previously written as a single byte. I capped the derived field numbers to 2047, which makes it a little easier to think about encoded_len and avoids thinking about reserved ranges or other limits.

defunctzombie commented on rust/foxglove/src/protobuf.rs:
```diff
@@ -151,6 +172,10 @@ impl ProtobufField for u8 {
     }
+
+    fn encoded_len(&self) -> usize {
+        prost::encoding::encoded_len_varint(*self as u64)
```
Is a u8 ever written as a varint?

bryfox replied:
Yes, and anything above 127 (highest bit set) will use two bytes. I added a test.

defunctzombie commented on rust/foxglove/src/protobuf.rs:
```diff
@@ -233,6 +282,10 @@ impl ProtobufField for bool {
     }
+
+    fn encoded_len(&self) -> usize {
+        prost::encoding::encoded_len_varint(*self as u64)
```
bool would always be 1 byte?

bryfox replied:
Yes, updated 👍

defunctzombie commented on rust/foxglove-derive/src/lib.rs:
```diff
@@ -68,6 +76,10 @@ fn derive_enum_impl(input: &DeriveInput, data: &DataEnum) -> TokenStream {
             }
+
+            fn encoded_len(&self) -> usize {
+                ::foxglove::protobuf::encoded_len_varint(*self as u64)
```
Why does this cast self to u64?

bryfox replied:
The cast came from the original implementation of `write` in the base branch, which was basically doing the same thing as prost, and prost's interface accepts u64. It works for the standard varint encoding because both [protobuf](https://protobuf.dev/programming-guides/encoding/#signed-ints) and [rust](https://doc.rust-lang.org/book/ch03-02-data-types.html#integer-types) use two's complement for signed ints. Should we do something more explicit here?

gasmith commented:
> https://github.com/foxglove/foxglove-sdk/blob/main/rust/foxglove/src/encode.rs#L31-L36
> 
> The comments here make this optional but it seems like you are saying it should not be optional? or that it is important to provide this?

It should probably be mandatory. The `Channel::log` implementation [panics](https://github.com/foxglove/foxglove-sdk/blob/3f2fcd2e84f9d32ea37ae549e4eba780ef96d9e7/rust/foxglove/src/channel.rs#L132-L138) for any message that requires more than 256k bytes, but does not implement `Encode::encoded_len()`.

eloff commented:
> > https://github.com/foxglove/foxglove-sdk/blob/main/rust/foxglove/src/encode.rs#L31-L36
> 
> > 
> 
> > The comments here make this optional but it seems like you are saying it should not be optional? or that it is important to provide this?
> 
> 
> 
> It probably should be mandatory, or at least the documentation should be clear about the situations in which it is mandatory. The `Channel::log` implementation [panics](https://github.com/foxglove/foxglove-sdk/blob/3f2fcd2e84f9d32ea37ae549e4eba780ef96d9e7/rust/foxglove/src/channel.rs#L132-L138) for any message that requires more than 256k bytes, but does not implement `Encode::encoded_len()`.

That's probably a limitation that could be lifted.

gasmith commented:
> > It probably should be mandatory, or at least the documentation should be clear about the situations in which it is mandatory. The `Channel::log` implementation [panics](https://github.com/foxglove/foxglove-sdk/blob/3f2fcd2e84f9d32ea37ae549e4eba780ef96d9e7/rust/foxglove/src/channel.rs#L132-L138) for any message that requires more than 256k bytes, but does not implement `Encode::encoded_len()`.
> 
> That's probably a limitation that could be lifted. I could reuse the Arena I implemented for the c++ schemas PR, I'll make the change there.

I misspoke, actually. It looks like `SmallBytes` implementation of `BufMut` does resize the buffer dynamically. I don't know why I thought otherwise.

eloff commented:
> > > It probably should be mandatory, or at least the documentation should be clear about the situations in which it is mandatory. The `Channel::log` implementation [panics](https://github.com/foxglove/foxglove-sdk/blob/3f2fcd2e84f9d32ea37ae549e4eba780ef96d9e7/rust/foxglove/src/channel.rs#L132-L138) for any message that requires more than 256k bytes, but does not implement `Encode::encoded_len()`.
> > 
> > 
> > That's probably a limitation that could be lifted. I could reuse the Arena I implemented for the c++ schemas PR, I'll make the change there.
> 
> I misspoke, actually. It looks like `SmallBytes` implementation of `BufMut` does resize the buffer dynamically. I don't know why I thought otherwise.

I came here to say this, I looked into the code, and it handles the case just fine, using encoded_len as an optimization hint, but not requiring that it's accurate.

So all we need to do is update the comment to specify that it's optional, but providing a good estimate, and erring on the side of over-estimating, is important to serialization performance.

bryfox commented on rust/foxglove/src/protobuf.rs:
```diff
@@ -219,6 +262,12 @@ impl ProtobufField for i8 {
+    fn encoded_len(&self) -> usize {
+        let n = *self;
+        let encoded = ((n << 1) ^ (n >> 7)) as u64;
+        prost::encoding::encoded_len_varint(encoded)
```
Yes, up to 10 bytes. This is currently using [zig-zag encoding](https://protobuf.dev/programming-guides/encoding/#signed-ints) (documented here as "Signed integers are encoded using sint32 or sint64."). An alternative is the standard/two's complement encoding, where every negative number takes up 10 bytes. I added a test here too.

eloff replied:
Shouldn't a zig-zag encoded i8 be at most 2 bytes? (only when you need 8 bits to represent the encoded integer)

eloff replied:
```suggestion
        if encoded < 128 { 1 } else { 2 }
```

I think you could even avoid doing the zig-zag encoding by checking n >= -64 && n < 64 (double check I don't have an off-by-one error there)

bryfox replied:
> Shouldn't a zig-zag encoded i8 be at most 2 bytes?

Yes, thank you! I had confused myself in some way. The encoding was wrong here, as was my subsequent test, since the ints weren't wide enough for the shift. Fixed in 215345c and added more tests.

> having a consistent, readable implementation is probably better

I think I agree with this, but I'm happy to file an issue to benchmark and implement if you think that's worthwhile.

bryfox commented:
[d03e1ea](https://github.com/foxglove/foxglove-sdk/pull/408/commits/d03e1ea8e11cf7454142b48ffc7cbaf51fea01d3) fixes the tag encoding (see updated description). I think I've addressed feedback above. The remaining issue is to solve the crate deployment, which I'll deal with in a separate PR.
